File: web-app/app/js/modules/balancePredictor.js
import { sum, monthKey as toMonthKey } from './utils.js';

const parseMonthKey = (mk) => {
  if (typeof mk !== 'string') return null;
  const match = mk.match(/^(\d{4})-(\d{2})$/);
  if (!match) return null;
  const year = Number(match[1]);
  const month = Number(match[2]);
  if (!Number.isFinite(year) || !Number.isFinite(month) || month < 1 || month > 12) return null;
  return { year, month };
};

const dayFromDate = (date) => {
  if (typeof date !== 'string') return null;
  const parts = date.split('-');
  if (parts.length !== 3) return null;
  const day = Number(parts[2]);
  return Number.isFinite(day) ? day : null;
};

const daysInMonth = (year, month) => {
  if (!Number.isFinite(year) || !Number.isFinite(month)) return 31;
  return new Date(year, month, 0).getDate();
};

const accumulateByDay = (transactions, monthLength) => {
  const totals = Array(monthLength + 1).fill(0);
  for (const tx of transactions || []) {
    const day = dayFromDate(tx?.date);
    if (!day || day < 1 || day > monthLength) continue;
    const amount = Number(tx?.amount) || 0;
    totals[day] += amount;
  }
  const cumulative = Array(monthLength + 1).fill(0);
  for (let day = 1; day <= monthLength; day += 1) {
    cumulative[day] = cumulative[day - 1] + totals[day];
  }
  return { totals, cumulative };
};

const median = (values) => {
  if (!values || values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  if (sorted.length % 2) return sorted[mid];
  return (sorted[mid - 1] + sorted[mid]) / 2;
};

const computeHistoryRemainders = (months, excludeKey) => {
  const map = new Map();
  for (const [mk, month] of Object.entries(months || {})) {
    if (!month || mk === excludeKey) continue;
    const parsed = parseMonthKey(mk);
    if (!parsed) continue;
    const monthLength = daysInMonth(parsed.year, parsed.month);
    const { cumulative } = accumulateByDay(month.transactions || [], monthLength);
    const finalSpend = cumulative[monthLength] || 0;
    for (let day = 0; day <= monthLength; day += 1) {
      const spentSoFar = cumulative[day] || 0;
      const remainder = finalSpend - spentSoFar;
      if (!map.has(day)) map.set(day, []);
      map.get(day).push(remainder);
    }
  }
  return map;
};

const pickRemainder = (map, targetDay) => {
  if (!map || map.size === 0) return { remainder: 0, sourceDay: null, sampleSize: 0 };
  const direct = map.get(targetDay);
  if (direct && direct.length) {
    return { remainder: median(direct), sourceDay: targetDay, sampleSize: direct.length };
  }
  for (let offset = 1; offset < 32; offset += 1) {
    const lower = targetDay - offset;
    if (lower >= 0) {
      const arr = map.get(lower);
      if (arr && arr.length) {
        return { remainder: median(arr), sourceDay: lower, sampleSize: arr.length };
      }
    }
    const higher = targetDay + offset;
    const arr = map.get(higher);
    if (arr && arr.length) {
      return { remainder: median(arr), sourceDay: higher, sampleSize: arr.length };
    }
  }
  const entries = [...map.entries()].sort((a, b) => a[0] - b[0]);
  if (entries.length) {
    const [day, arr] = entries[entries.length - 1];
    return { remainder: median(arr), sourceDay: day, sampleSize: arr.length };
  }
  return { remainder: 0, sourceDay: null, sampleSize: 0 };
};

const determineObservationDay = (monthKey, transactions, today) => {
  const parsed = parseMonthKey(monthKey);
  if (!parsed) return { day: 0, monthLength: 31 };
  const monthLength = daysInMonth(parsed.year, parsed.month);
  const todayKey = toMonthKey(today);
  if (monthKey < todayKey) {
    return { day: monthLength, monthLength };
  }
  const txDays = (transactions || [])
    .map((tx) => dayFromDate(tx?.date))
    .filter((day) => day && day >= 1 && day <= monthLength);
  if (monthKey === todayKey) {
    const todayDay = Math.min(today.getDate(), monthLength);
    if (txDays.length) {
      return { day: Math.max(Math.max(...txDays), todayDay), monthLength };
    }
    return { day: todayDay, monthLength };
  }
  if (txDays.length) {
    return { day: Math.max(...txDays), monthLength };
  }
  return { day: 0, monthLength };
};

export const predictBalance = (monthKey, months, today = new Date()) => {
  if (!months || !monthKey) return null;
  const month = months[monthKey];
  if (!month) return null;
  const { day: observationDay, monthLength } = determineObservationDay(
    monthKey,
    month.transactions || [],
    today,
  );
  const { cumulative } = accumulateByDay(month.transactions || [], monthLength);
  const spentSoFar = cumulative[Math.min(observationDay, monthLength)] || 0;
  const incomesTotal = sum(month.incomes || [], (inc) => Number(inc?.amount) || 0);
  const remainders = computeHistoryRemainders(months, monthKey);
  const { remainder, sourceDay, sampleSize } = pickRemainder(remainders, observationDay);
  const predictedSpend = Math.max(0, spentSoFar + remainder);
  const predictedLeftover = incomesTotal - predictedSpend;
  return {
    predictedSpend,
    predictedLeftover,
    spentSoFar,
    incomesTotal,
    observationDay,
    remainderUsedDay: sourceDay,
    sampleSize,
  };
};

export const __private = {
  parseMonthKey,
  dayFromDate,
  daysInMonth,
  accumulateByDay,
  computeHistoryRemainders,
  pickRemainder,
  determineObservationDay,
  median,
};

File: android/app/src/main/java/com/homebudgeting/domain/BalancePredictor.kt
package com.homebudgeting.domain

import com.homebudgeting.data.BudgetMonth
import com.homebudgeting.data.BudgetTransaction
import java.time.LocalDate
import java.time.YearMonth
import kotlin.math.max

private data class RemainderStats(
    val remainder: Double,
    val sourceDay: Int?,
    val sampleSize: Int
)

data class BalancePrediction(
    val predictedSpend: Double,
    val predictedLeftover: Double,
    val spentSoFar: Double,
    val incomesTotal: Double,
    val observationDay: Int,
    val remainderUsedDay: Int?,
    val sampleSize: Int
)

fun predictBalance(
    monthKey: String?,
    months: Map<String, BudgetMonth>,
    today: LocalDate = LocalDate.now()
): BalancePrediction? {
    if (monthKey.isNullOrBlank()) return null
    val month = months[monthKey] ?: return null
    val observation = determineObservationDay(monthKey, month.transactions, today)
    val monthLength = observation.monthLength
    val cumulative = accumulateByDay(month.transactions, monthLength)
    val spentSoFar = cumulative[minOf(observation.day, monthLength)]
    val incomesTotal = month.incomes.sumOf { it.amount }
    val remainders = computeHistoryRemainders(months, monthKey)
    val remainderStats = pickRemainder(remainders, observation.day)
    val predictedSpend = max(0.0, spentSoFar + remainderStats.remainder)
    val predictedLeftover = incomesTotal - predictedSpend
    return BalancePrediction(
        predictedSpend = predictedSpend,
        predictedLeftover = predictedLeftover,
        spentSoFar = spentSoFar,
        incomesTotal = incomesTotal,
        observationDay = observation.day,
        remainderUsedDay = remainderStats.sourceDay,
        sampleSize = remainderStats.sampleSize
    )
}

private data class Observation(val day: Int, val monthLength: Int)

private fun determineObservationDay(
    monthKey: String,
    transactions: List<BudgetTransaction>,
    today: LocalDate
): Observation {
    val parsed = parseMonthKey(monthKey) ?: return Observation(day = 0, monthLength = 31)
    val monthLength = parsed.lengthOfMonth()
    val todayKey = YearMonth.from(today).toString()
    if (monthKey < todayKey) {
        return Observation(day = monthLength, monthLength = monthLength)
    }
    val validDays = transactions.mapNotNull { it.date.takeLast(2).toIntOrNull() }
        .filter { it in 1..monthLength }
    if (monthKey == todayKey) {
        val todayDay = minOf(today.dayOfMonth, monthLength)
        val observed = (validDays + todayDay).maxOrNull() ?: todayDay
        return Observation(observed, monthLength)
    }
    if (validDays.isNotEmpty()) {
        return Observation(validDays.maxOrNull() ?: 0, monthLength)
    }
    return Observation(day = 0, monthLength = monthLength)
}

private fun accumulateByDay(
    transactions: List<BudgetTransaction>,
    monthLength: Int
): DoubleArray {
    val totals = DoubleArray(monthLength + 1)
    transactions.forEach { tx ->
        val day = tx.date.takeLast(2).toIntOrNull()
        if (day != null && day in 1..monthLength) {
            totals[day] += tx.amount
        }
    }
    val cumulative = DoubleArray(monthLength + 1)
    for (day in 1..monthLength) {
        cumulative[day] = cumulative[day - 1] + totals[day]
    }
    return cumulative
}

private fun computeHistoryRemainders(
    months: Map<String, BudgetMonth>,
    excludeKey: String
): Map<Int, MutableList<Double>> {
    val map = mutableMapOf<Int, MutableList<Double>>()
    months.forEach { (key, month) ->
        if (key == excludeKey) return@forEach
        val parsed = parseMonthKey(key) ?: return@forEach
        val length = parsed.lengthOfMonth()
        val cumulative = accumulateByDay(month.transactions, length)
        val finalSpend = cumulative[length]
        for (day in 0..length) {
            val spent = cumulative[day]
            val remainder = finalSpend - spent
            map.getOrPut(day) { mutableListOf() }.add(remainder)
        }
    }
    return map
}

private fun pickRemainder(
    map: Map<Int, MutableList<Double>>,
    targetDay: Int
): RemainderStats {
    if (map.isEmpty()) return RemainderStats(0.0, null, 0)
    map[targetDay]?.takeIf { it.isNotEmpty() }?.let {
        return RemainderStats(it.median(), targetDay, it.size)
    }
    for (offset in 1..31) {
        val lower = targetDay - offset
        if (lower >= 0) {
            map[lower]?.takeIf { it.isNotEmpty() }?.let {
                return RemainderStats(it.median(), lower, it.size)
            }
        }
        val higher = targetDay + offset
        map[higher]?.takeIf { it.isNotEmpty() }?.let {
            return RemainderStats(it.median(), higher, it.size)
        }
    }
    val lastEntry = map.entries.maxByOrNull { it.key }
    return if (lastEntry != null) {
        RemainderStats(lastEntry.value.median(), lastEntry.key, lastEntry.value.size)
    } else {
        RemainderStats(0.0, null, 0)
    }
}

private fun List<Double>.median(): Double {
    if (isEmpty()) return 0.0
    val sorted = this.sorted()
    val mid = size / 2
    return if (size % 2 == 0) {
        (sorted[mid - 1] + sorted[mid]) / 2.0
    } else {
        sorted[mid]
    }
}

private fun parseMonthKey(key: String): YearMonth? = runCatching {
    YearMonth.parse(key)
}.getOrNull()
